### ISSUE TITLE
Switch expression is not exhaustive on integers

### ISSUE THREAD

**Issue by Pante** [dart-lang/language#3083]
* Dart 3.0.0

The analyzer claims that the following code snippet isn't exhaustive when it is. It emits the following message: `The type 'int' is not exhaustively matched by the switch cases since it doesn't match 'int()'`.

```dart
void main() {
  final foo = switch ('a'.compareTo('b')) {
    > 0 => 'more',
    == 0 => 'equal',
    < 0 => 'less',
  };
}
```

**Comment by lrhn** [comment]
The first switch is not exhaustive because there is no specified way to exhaust integers.
Exhausting integers by range exhaustion, like this example tries, is something that *could* be possible (probably), with some possible caveats about the difference between web numbers and native integers.
But it's not part of the current exhaustiveness definition.
See https://github.com/dart-lang/language/issues/2474#issuecomment-1244099229
(Don't know if we have an open issue for it.)

The second example doesn't work because a case with a `with` clause never counts towards exhaustiveness. The compiler cannot be expected to evaluate arbitrary expressions for all possible values while compiling, so if a `case` has a `when`, the compiler goes "damn if I know what that means", and ignores it for exhaustion checking.
Which means that. as far as the compiler knows, the example switch can reach the point after the two first cases with any combination of values still possible. That's what the error message says: It (possibly) doesn't match *anything*.
That's working as intended, and is unlikely to change.

It's not because `(bool?, bool?)` is not sealed/enum, it's treated just as if `bool` was an enum, and so the entire record is a "must exhaust" type, and it _can_ be exhausted by, e.g.,
```dart
bool bothTheSame(bool? a, bool? b) => switch ((a, b)) {
    (true, true) => true,
    (true, false) => false,
    (true, null) => false,
    (false, false) => true,
    (false, true) => false,
    (false, null) => false,
    (null, null) => true,
    (null, true) => false,
    (null, false) => false,
 };
```

(But that's just doing `bool bothTheSame(bool? a, bool? b) => a == b;` the long way 😉.)

**Comment by leafpetersen** [comment]
Or more succinctly (though not as succinct as `a == b`):

```dart
bool bothTheSame(bool? a, bool? b) => switch ((a, b)) {
    (true, true) => true,
    (false, false) => true,
    (null, null) => true,
    (_, _) => false
 };
```

**Comment by munificent** [comment]
We definitely could do exhaustiveness checking over integer ranges with relational patterns. It's something some other languages do and is something I'd like to do for Dart. We just didn't have time to get it in to 3.0.

**Comment by adriank** [comment]
@munificent could that also be possible for double?

**Comment by lrhn** [comment]
No, you cannot exhaust doubles using `<`, `==` and `>` patterns, or any non-match-all pattern, because you cannot match NaN using any of the match types which work towards exhaustiveness.
(Getting more an more convinced that a constant pattern of `double.nan` should match any NaN value, rather than use `==` and give false. Because then it's actually useful instead of completely useless. Won't be able to distinguish different NaN values that way, but who cares.)

**Comment by MelbourneDeveloper** [comment]
Generally not a huge deal, but when this bites you, it bites you hard because it's very confusing to figure out why this is happening. If you don't know that the exhaustiveness checker doesn't handle integers like this, you'll be scratching your head.

**Comment by rfitz123** [comment]
I don't love having to add an untestable line of code to use relational pattern matching. It's an otherwise very convenient feature.

**Comment by lrhn** [comment]
The biggest issue I see is that integers are not working the same way in native and web code. On the web, `double.inifinity` is an integer.
The exhaustiveness checking would have to be general enough to cover both number implementations, which means it likely has to use `BigInt`s for the numbers, and not assume a max or min value, and not assume any holes in the integers.
(So it should work on generalized mathematical integers, not Dart representations.)

Then it can probably be made to work.

**Comment by mmcdon20** [comment]
> The biggest issue I see is that integers are not working the same way in native and web code. On the web, `double.inifinity` is an integer. 

Wouldn't that still get matched by the `case > 0:`?

 `<0`, `>0`, and `==0` would still appear to exhaust all cases even with the possibility of infinity values.

**Comment by munificent** [comment]
> `<0`, `>0`, and `==0` would still appear to exhaust all cases even with the possibility of infinity values.

They won't cover `NaN`.

**Comment by lrhn** [comment]
Sure, every integer would be mathced by `< 0 || == 0 || > 0` (but not doubles, because NaN).
That's the easy case, and most likely 99.999% of what people will actually write is easy cases.

The language specification has to specify what happens in all cases, in a consistent way, and *preferably* in a way that is not platform dependent. Otherwise the analyzer might tell you that your switch is exhaustive, but when you compile it for the web, it breaks.

So, which of these switch expressions will compile, on which platforms, and would any change if you wrote an integer literal in place of a named constant?
```dart
void checkIt(int value) {
   /*01*/ print(switch (value) { < 0 || == 0 || > 0 => "ok" });
   /*02*/ print(switch (value) { >= int.minValue => "ok" });
   /*03*/ print(switch (value) { <= int.maxValue => "ok" });
   /*04*/ print(switch (value) { >= -0x8000_0000_0000_0000 => "ok" });
   /*05*/ print(switch (value) { <= (0x8000_0000_0000_0000 - 1) => "ok" });
   /*06*/ print(switch (value) { < -0.0 || == 0.0 || > 0.0 => "ok" });  // Operators accept doubles.
   /*07*/ print(switch (value) { < 0.5 || > 0.5 => "ok" });
   /*08*/ print(switch (value) { != 0.5 => "ok" });

   /*09*/ print(switch (value) { <= i2_53 || >= i2_53n => "ok" });
   /*10*/ print(switch (value) { < i2_53n || > i2_53n => "ok" });
   /*11*/ print(switch (value) { <= i2_53n || >= i2_53n2 => "ok" });
   /*12*/ print(switch (value) { <= d2_53 || >= d2_53n => "ok" });
   /*13*/ print(switch (value) { < i2_53n || > i2_53n => "ok" });
   /*14*/ print(switch (value) { <= i2_53n || >= d2_53n => "ok" });

   /*15*/ print(switch (value) { <= i2_63p => "ok" });
   /*16*/ print(switch (value) { <= d2_63p || >= d2_63 => "ok" });
   /*17*/ print(switch (value) { <= d2_63 || >= d2_63n => "ok" });
   /*18*/ print(switch (value) { <= i2_63p || >= d2_63 => "ok" });
   /*19*/ print(switch (value) { <= i2_63 || >= d2_63n => "ok" });
   /*20*/ print(switch (value) { <= d2_63p || >= i2_63p => "ok" });

   /*21*/ print(switch (value) { < d2_64 => "ok" });
   /*22*/ print(switch (value) { <= d2_64 || >= d2_64n => "ok" });

   /*23*/ print(switch (value) { == 0 || != i2_64 => "ok" });
}

// Constant variables' values
//         : Native                 : Web
// i2_32   :             4294967296 :           4294967296
// i2_53   :       9007199254740992 :     9007199254740992
// i2_53n  :       9007199254740993 :     9007199254740992
// i2_53n2 :       9007199254740994 :     9007199254740994
// d2_53   :     9007199254740992.0 :     9007199254740992
// d2_53n  :     9007199254740994.0 :     9007199254740994
// in2_63  :   -9223372036854775808 : -9223372036854775808
// i2_63   :   -9223372036854775808 :  9223372036854775808
// i2_63p  :    9223372036854775807 :  9223372036854775808
// d2_63   :  9223372036854775808.0 :  9223372036854775808
// d2_63p  :  9223372036854774784.0 :  9223372036854774784
// d2_63n  :  9223372036854777856.0 :  9223372036854777856
// i2_64   :                      0 : 18446744073709551616
// d2_64   : 18446744073709551616.0 : 18446744073709551616
// d2_64n  : 18446744073709555712.0 : 18446744073709555712

const int i2_32 = 0x1_0000_0000; // int 2^32
const int i2_53 = 0x20_0000_0000_0000; // int 2^53
const int i2_53n = i2_53 + 1; /// Next integer after i2_53
const int i2_53n2 = i2_53 + 2; // next integer after i2_53n
const double d2_53 = 0x20_0000_0000_0000; // double 2^53 (double.maxPrecise)
const double d2_53n = d2_53 + 2; // next double after d2_53.
const int in2_63 = -0x8000_0000_0000_0000; // -2^63, int.minValue,
const int i2_63 = -in2_63; // 2^63, overflows to -2^63 on native, not on web.
const int i2_63p = 0x8000_0000_0000_0000 - 1; // 2^63 - 1. Not precise on web.
const double d2_63 = 0x8000_0000_0000_0000; // double 2^63.
const double d2_63p = d2_63 - 0x400; // prev double before 2^63.
const double d2_63n = d2_63 + 0x800; // next double after 2^63
const int i2_64 = i2_32 * i2_32; // Overflows to zero on native, not on web.
const double d2_64 = 0x1_0000_0000_0000_0000; // double 2^64
const double d2_64n = d2_64 + 0x1000; // 2^64+2^12, next double after d2_64.
```

To answer those, there needs to be a well-defined strategy for deciding which ranges of which integers are covered by which tests.
The *constants themselves* can vary between platforms, because they are computed using the platform's number semantics. If exhaustiveness of integers should not change between platforms, we may need to step outside of the platform numbers themselves.

Maybe:
> Integer numerals written in relational operators for `int` matched values are not interpreted as integer constants, but as "mathematical integers", only such comparisons can actally exhaust `int`, and then they need to cover the entire mathematical set if integers, ℤ.

Because then the constants themselves would not be subject to platform-specific interpretation, they'd either cover ℤ or not, and that would be true whether `int` is represented as int-64 or float-64. 

That means `switch (someInt) { < 0 || > 0 => ... , == 0 => ... }` will exhaust integers, but `const z = 0; ... switch (someInt) { < z || > z => ..., == 0 => ...}` will not.
That feels odd, but if the declaration of `z` was `const z = identical(1, 1.0) ? 1 : 0;`, then the *constant value* could vary between platforms, and it would not be exhaustive on web. So the switch is not statically known to be exhaustive, just because it happens to be for *this* platform's constant values.

It's very special-casey, but integers _are_ very special.

(We can have platform-specific constant problems in other places too. We have disallowed values depending on `String.fromEnvironment`-and-friends in switches, but that's not the only way to have different compilations.
```dart
enum E {
  v1, v2; 
  static const v0 = (identical(1, 1.0) ? E.v1 : E.v2); 
}
int check(E e) => switch (e) { E.v0 => 0, E.v1 => 1 };
```
This code is exhaustive on native, not on web, because it's not the same switch cases on different platforms.
That's just a weird thing to do, so most people don't, but differences in number semantics come up much more often.)
  

TL;DR: _Constants aren't._
